feat: Add auth token list and delete commands

lhsazevedo opened the pull request:
:wave: Hi, this is my first contribution :)

* Resolves #39690
* Resolves #39691

## Summary
Add two new commands requested in #39690 and #39691.

Since both commands can handle app passwords as well as session tokens, I've named them `occ user:auth-tokens` and `occ user:delete-auth-token` instead to better align with the nouns used in the codebase. Let me know if the requested names should be kept.

## Sample output:
```
$ ./occ user:auth-tokens admin
  - 0:
    - id: 8
    - name: Mozilla/5.0 (X11; Linux x86_64; rv:107.0) Gecko/20100101 Firefox/107.0
    - lastActivity: 1692835829
    - type: 0
    - scope:
      - filesystem: true
  - 1:
    - id: 9
    - name: my-app-password
    - lastActivity: 1692835843
    - type: 1
    - scope:
      - filesystem: true

$ ./occ user:delete-auth-token 9
$ ./occ user:auth-tokens --output=json_pretty admin
[
    {
        "id": 8,
        "name": "Mozilla\/5.0 (X11; Linux x86_64; rv:107.0) Gecko\/20100101 Firefox\/107.0",
        "lastActivity": 1692845431,
        "type": 0,
        "scope": {
            "filesystem": true
        }
    }
]
```

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

lhsazevedo committed:
feat: Add user:auth-tokens command

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo committed:
feat: Add user:delete-auth-token command

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo committed:
Add missing return types

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo commented:
I'll open the documentation PR as soon as we agree on this one!

Also, while implementing the commands I noticed some possible minor follow ups:

1. #39691 also mentions the possibility of deleting tokens older than a given timestamp or age. I think this could be done is by adding `--before` and `--age` options, either in the introduced `delete-auth-token` command (by making `<id>` optional) or in a new one to avoid confusion. 
2. `OC\Core\Command\Base::configure()`:
   1. A few commands are manually setting up the `--output` option instead of relying on the parent `OC\Core\Command\Base::configure()` method for this.
      - core/Command/Group/Info.php
      - core/Command/Group/ListCommand.php
      - core/Command/User/Info.php
      - core/Command/User/ListCommand.php
   2. Despite not allowing control over the output format, some commands have the `--output` option because they are calling `parent::configure()`:
      - core/Command/App/GetPath.php
      - core/Command/TwoFactorAuth/Enable.php
      - core/Command/TwoFactorAuth/Disable.php
      - core/Command/TwoFactorAuth/Cleanup.php
      - core/Command/TwoFactorAuth/State.php
      - core/Command/Config/App/DeleteConfig.php
      - core/Command/Config/App/SetConfig.php
      - core/Command/Config/System/DeleteConfig.php
      - core/Command/Config/System/SetConfig.php
   
   I guess that both of the above happened because it isn't clear that `OC\Core\Command\Base::configure()` only sets up the `output` option. What do you think of renaming it to `setupOutputFormatOption()`?.
3. There now three auth token related commands in `core/Commands/User`. Maybe we could group them inside a nested `AuthToken` namespace?

Is there anything in the list that you think should be done?

nickvergessen commented:
> 1. https://github.com/nextcloud/server/issues/39691 also mentions the possibility of deleting tokens older than a given timestamp or age. I think this could be done is by adding --before and --age options, either in the introduced delete-auth-token command (by making <id> optional) or in a new one to avoid confusion.

I would make id optional and go with a `--before` parameter as you suggested

2. `configure()` is a "magic" predefined function of the Symfony console, so would like to keep that (also it would break apps that use it). Calling `parent::configure()` is the prefered way and if the used options are the same, those not using it should adjusted in the future.

3. Makes sense to me

nickvergessen commented on core/Command/User/AuthTokens.php:
```diff
@@ -0,0 +1,70 @@
+		parent::configure();
+
+		$this
+			->setName('user:auth-tokens')
```
How about namespacing them as well?
```suggestion
			->setName('user:auth-tokens:list')
```

and rename the others to
```
			->setName('user:auth-tokens:delete')
```
```
			->setName('user:auth-tokens:add')
			->setAliases(['user:add-app-password'])
```

lhsazevedo replied:
Sounds great, the alias was a good idea as well. Done.

nickvergessen commented on core/Command/User/AuthTokens.php:
```diff
@@ -0,0 +1,70 @@
+
+		$tokens = $this->tokenProvider->getTokenByUser($user->getUID());
+
+		$data = array_map(fn (IToken $token): mixed => $token->jsonSerialize(), $tokens);
```
I think we should filter out some columns here:
- password
- password_hash
- token
- public_key
- private_key

lhsazevedo replied:
Done. For the record, `PublicKeyToken::jsonSerialize()` already explicitly lists allowed fields, ao we are adding a second guard here to be on the safe side.

nickvergessen commented on core/register_command.php:
```diff
@@ -193,6 +193,8 @@
 	$application->add(new OC\Core\Command\User\ListCommand(\OC::$server->getUserManager(), \OC::$server->getGroupManager()));
 	$application->add(new OC\Core\Command\User\Info(\OC::$server->getUserManager(), \OC::$server->getGroupManager()));
 	$application->add(new OC\Core\Command\User\AddAppPassword(\OC::$server->get(\OCP\IUserManager::class), \OC::$server->get(\OC\Authentication\Token\IProvider::class), \OC::$server->get(\OCP\Security\ISecureRandom::class), \OC::$server->get(\OCP\EventDispatcher\IEventDispatcher::class)));
+	$application->add(new OC\Core\Command\User\AuthTokens(\OC::$server->get(\OCP\IUserManager::class), \OC::$server->get(\OC\Authentication\Token\IProvider::class)));
+	$application->add(new OC\Core\Command\User\DeleteAuthToken(\OC::$server->get(\OC\Authentication\Token\IProvider::class)));
```
Since all parameters are DI-able we can also just use:
```suggestion
	$application->add(\OC::$server->get(\OC\Core\Command\User\AuthTokens::class));
	$application->add(\OC::$server->get(\OC\Core\Command\User\DeleteAuthToken::class));
```

lhsazevedo replied:
Oh I see, nice

nickvergessen commented:
> I would make id optional and go with a --before parameter as you suggested

That being said, with `--before` you definetly want to limit it to a user. So maybe the delete should take the userId as first argument as some insurance if you delete by id and to allow `--before` being scoped to a user?

nickvergessen commented:
Otherwise that's a nice and handy PR :)

nickvergessen commented on core/Command/User/AuthTokens.php:
To fix CI:
> Please run: bash build/autoloaderchecker.sh
> And commit the result

lhsazevedo replied:
Thanks

tcitworld commented on core/Command/User/AuthTokens.php:
```diff
@@ -0,0 +1,70 @@
+
+		$data = array_map(fn (IToken $token): mixed => $token->jsonSerialize(), $tokens);
+
+		$this->writeArrayInOutputFormat($input, $output, $data);
```
**suggestion:** Maybe `writeTableInOutputFormat` is more suited?

lhsazevedo replied:
The only block I found here is that the scope field is an array (decoded JSON), and `writeTableInOutputFormat` expects fields to be scalars (or objects implementing `__toString()`). I was only able to find references to the `filesystem` scope, but there could be more in the future and I'm not sure if it would fit nicely in a table cell. We might as well omit the scope field when using the plain output format (table), or proceed with the simpler `writeArrayInOutputFormat`...

What do you think?

tcitworld replied:
As long as the array stays kinda the same we're safe.

A possibility would be to `array_filter` the array for `true` values, get only the keys (`array_keys`) from what's left and `implode` that to a string.

In any case, adding a test on the output should make sure format changes don't break anything.

lhsazevedo replied:
Here's the result

![image](https://github.com/nextcloud/server/assets/7695608/76b57fbc-e4e6-4110-8c80-bf8acd62b1c7)
~chef's kiss

This is done only when the output format is `OUTPUT_FORMAT_PLAIN`. Otherwise, we keep the array to be consumed by scripts.

tcitworld replied:
@lhsazevedo Could you format `lastActivity` to an ISO8601 date string and map [the `type` constants](https://github.com/nextcloud/server/blob/b570d76355c9c463c7ded49ac384a2aaa5ba36a4/lib/private/Authentication/Token/IToken.php#L32-L34C20) to strings?

lhsazevedo replied:
Sure

lhsazevedo replied:
@tcitworld Done, I took the opportunity to extract the plain output code to a method.

```
$ occ user:auth-tokens:list --no-ansi admin
+----+------------------------------------------------------------------------+---------------------------+-----------+------------+
| id | name                                                                   | lastActivity              | type      | scope      |
+----+------------------------------------------------------------------------+---------------------------+-----------+------------+
| 3  | cli                                                                    | 2023-08-28T01:32:34+00:00 | permanent | filesystem |
| 6  | Mozilla/5.0 (X11; Linux x86_64; rv:107.0) Gecko/20100101 Firefox/107.0 | 2023-08-28T01:47:08+00:00 | temporary | filesystem |
| 7  | my-test-app                                                            | 2023-08-28T01:47:40+00:00 | permanent | filesystem |
+----+------------------------------------------------------------------------+---------------------------+-----------+------------+
```

lhsazevedo committed:
Filter out sensitive fields in user:auth-tokens

PublicKeyToken::jsonSerialize() already explicitly lists allowed fields,
we are adding a second guard here to be on the safe side.

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo committed:
Use autowiring when registering commands

Co-authored-by: Joas Schilling <213943+nickvergessen@users.noreply.github.com>
Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo committed:
Namespace user auth token commands

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo committed:
Update autoloaders

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

nickvergessen commented:
Let's use `--last-used-before`, then it should be clear and we can add the other one later if needed?

lhsazevedo committed:
Add last-used-before option

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo committed:
Add tests for occ user:auth-tokens:delete

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo commented:
Done, thanks for reviewing! I could squash all the commits but that might make reviewing harder, so feel free to squash when merging...

nickvergessen commented on core/Command/User/AuthTokens/Delete.php:
```diff
@@ -0,0 +1,120 @@
+
+	protected function execute(InputInterface $input, OutputInterface $output): int {
+		$uid = $input->getArgument('uid');
+		$id = $input->getArgument('id');
```
```suggestion
		$id = (int) $input->getArgument('id');
```

nickvergessen commented on core/Command/User/AuthTokens/Delete.php:
```diff
@@ -0,0 +1,120 @@
+		return Command::SUCCESS;
+	}
+
+	protected function deleteLastUsedBefore(string $uid, string $before) {
```
```suggestion
	protected function deleteLastUsedBefore(string $uid, string $before): int {
```

nickvergessen commented on lib/private/Authentication/Token/PublicKeyTokenMapper.php:
```diff
@@ -69,6 +69,15 @@ public function invalidateOld(int $olderThan, int $remember = IToken::DO_NOT_REM
 
+	public function invalidateLastUsedBefore(string $uid, int $before): int {
+		$qb = $this->db->getQueryBuilder();
+		return $qb->delete($this->tableName)
+			->where($qb->expr()->eq('uid', $qb->createNamedParameter($uid)))
+			->andWhere($qb->expr()->lt('last_activity', $qb->createNamedParameter($before, IQueryBuilder::PARAM_INT)))
+			->andWhere($qb->expr()->eq('version', $qb->createNamedParameter(PublicKeyToken::VERSION, IQueryBuilder::PARAM_INT)))
+			->executeStatement();
```
```suggestion
		$qb->delete($this->tableName)
			->where($qb->expr()->eq('uid', $qb->createNamedParameter($uid)))
			->andWhere($qb->expr()->lt('last_activity', $qb->createNamedParameter($before, IQueryBuilder::PARAM_INT)))
			->andWhere($qb->expr()->eq('version', $qb->createNamedParameter(PublicKeyToken::VERSION, IQueryBuilder::PARAM_INT)));
		return $query->executeStatement();
```

Let's split type changing calls to a new expression

nickvergessen reviewed (commented):
Almost there, let's make Psalm a bit more happy

nickvergessen commented:
> I could squash all the commits but that might make reviewing a little harder, feel free to do so when merging.

Squashing after reviews are done is what we mostly do 👍🏼

lhsazevedo committed:
Fixes from static analysis

Co-authored-by: Joas Schilling <213943+nickvergessen@users.noreply.github.com>
Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo committed:
Fix typo

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo committed:
Use table output for list command

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo committed:
Format lastActivity and type for plain output

Signed-off-by: Lucas Azevedo <lhs_azevedo@hotmail.com>

lhsazevedo committed:
Merge branch 'master' into auth-token-commands

lhsazevedo commented:
Conflict resolved

nickvergessen commented:
Awesome story @lhsazevedo I invited you to the org

lhsazevedo commented:
Thanks for inviting me @nickvergessen, just joined it. I'm looking forward to contribute more with the project.
